### PR TITLE
more options in Kelly error estimator

### PR DESCRIPTION
**Pull request by davydden**
Kelly error estimator is extended to support
```
\eta^2 = \sum_face h_face / 2 p_e \int ...
where p_e = max(p_elem, p_neighbor);
```
I am not sure how one treats faces with hanging nodes. I have not seen it mentioned in the literature. 
So far, a value of integral for the parent face is set to a sum of factored integrals of child faces.
Please, let me know if you think it should be different.

For Neumann BC it would be
```
  h_face/ p_cell  \int ...
```

p.s. documentation has not been updated yet.

### CODE REVIEW

File: include/deal.II/numerics/error_estimator.h [hunk 1]
```diff
@@ -308,7 +314,8 @@ class KellyErrorEstimator
                         const Function<spacedim>     *coefficients   = 0,
                         const unsigned int       n_threads = numbers::invalid_unsigned_int,
                         const types::subdomain_id subdomain_id = numbers::invalid_subdomain_id,
-                        const types::material_id       material_id = numbers::invalid_material_id);
+                        const types::material_id       material_id = numbers::invalid_material_id,
+                        KellyErrorEstimator<dim,spacedim>::Strategy strategy = Ainsworth);
```
Comment (bangerth) [inline include/deal.II/numerics/error_estimator.h]: make the argument const to indicate that it is an input parameter.

File: include/deal.II/numerics/error_estimator.h [hunk 2]
```diff
@@ -324,7 +331,8 @@ class KellyErrorEstimator
...
```
Comment (bangerth) [inline include/deal.II/numerics/error_estimator.h]: same here and in all of the other places.

File: source/numerics/error_estimator.cc [hunk 3]
```diff
@@ -464,6 +464,98 @@ namespace internal
       return face_integral;
     }
 
+    /**
+     * A factor to scale the integral for the face at the boundary.
+     * Used for Neumann BC (?).
```
Comment (bangerth) [inline source/numerics/error_estimator.cc]: what does the question mark indicate?

File: source/numerics/error_estimator.cc [hunk 4]
```diff
@@ -464,6 +464,98 @@ namespace internal
...
+                                        typename KellyErrorEstimator<DH::dimension,DH::space_dimension>::Strategy strategy)
+    {
+      switch (strategy)
+        {
+        case KellyErrorEstimator<DH::dimension,DH::space_dimension>::Ainsworth :
+          return 1.0;
+          break;
+        case KellyErrorEstimator<DH::dimension,DH::space_dimension>::face_diameter_over_twice_max_degree :
+          const double cell_degree = fe_face_values_cell.get_fe_collection()[cell->active_fe_index()].degree;
+          return cell->face(face_no)->diameter() / cell_degree;
```
Comment (bangerth) [inline source/numerics/error_estimator.cc]: what degree do you take if this and the neighboring face have different polynomial degrees?

File: source/numerics/error_estimator.cc [hunk 5]
```diff
@@ -464,6 +464,98 @@ namespace internal
...
+    double regular_face_factor_boundary(const typename DH::active_cell_iterator &cell,
+                                        const unsigned int                       face_no,
+                                        const dealii::hp::FEFaceValues<DH::dimension, DH::space_dimension> &fe_face_values_cell,
+                                        typename KellyErrorEstimator<DH::dimension,DH::space_dimension>::Strategy strategy)
+    {
+      switch (strategy)
+        {
+        case KellyErrorEstimator<DH::dimension,DH::space_dimension>::Ainsworth :
+          return 1.0;
+          break;
```
Comment (bangerth) [inline source/numerics/error_estimator.cc]: here and below, no need to have a `break` after a `return` :-)

File: source/numerics/error_estimator.cc (hunk 4 above)
Comment (bangerth) [inline source/numerics/error_estimator.cc]: oh, oh, I see now, just a few lines below.

**Comment by bangerth** [comment]
Looks good in principle. Please add the documentation and a test (plus an entry in changes.h).

File: source/numerics/error_estimator.cc (hunk 3 above)
Comment (davydden) [inline source/numerics/error_estimator.cc]: i was not sure about this comment initially, but that's what indeed happens, i.e. `integrate_over_regular_face()` for `face->at_boundary()` is called for faces in Neumann part of the boundary.

**Comment by davydden** [comment]
will do that...

**Comment by davydden** [comment]
@bangerth i think it's ready. Please, give it a final look. If all is good, i'll stash

File: doc/news/changes.h [hunk 6]
```diff
@@ -43,7 +43,7 @@ inconvenience this causes.
   but the scalar type of the vector was always <code>double</code>. This
   has been changed so that the two always need to match.
   <br>
-  (David Davydov, Wolfgang Bangerth, 2015/06/17)
+  (Denis Davydov, Wolfgang Bangerth, 2015/06/17)
```
Comment (bangerth) [inline doc/news/changes.h]: Aw, that's embarrassing. My apologies :-(

**Comment by davydden** [comment]
p.s. `deal.II/error_estimator*` pass without issues for me.

File: doc/news/changes.h (hunk 6 above)
Comment (davydden) [inline doc/news/changes.h]: no worries

File: include/deal.II/numerics/error_estimator.h [hunk 7]
```diff
@@ -83,22 +92,23 @@ namespace hp
...
  * $[\cdot]$ denotes the jump of the argument at the face. In the paper of
- * Ainsworth, $h$ is divided by $24$, but this factor is a bit esoteric,
+ * Ainsworth $ c_F=\frac h{24} $, but this factor is a bit esoteric,
  * stemming from interpolation estimates and stability constants which may
  * hold for the Poisson problem, but may not hold for more general situations.
- * In the implementation, this factor is considered, but may lead to wrong
- * results. You may scale the vector appropriately afterwards.
+ * Alternatively, we consider the case when $ c_F=\frac {h_F}_{2p_F} $,
+ * where $ h_F $ is face diagonal and $ p_F=max(p^+,p^-) $ is the
+ * maximum polynomial degree of adjacent elements.
```
Comment (bangerth) [inline include/deal.II/numerics/error_estimator.h]: can you point out here how one would choose between the two?

File: include/deal.II/numerics/error_estimator.h [hunk 8]
```diff
@@ -308,7 +333,8 @@ class KellyErrorEstimator
                         const Function<spacedim>     *coefficients   = 0,
                         const unsigned int       n_threads = numbers::invalid_unsigned_int,
                         const types::subdomain_id subdomain_id = numbers::invalid_subdomain_id,
-                        const types::material_id       material_id = numbers::invalid_material_id);
+                        const types::material_id       material_id = numbers::invalid_material_id,
+                        const KellyErrorEstimator<dim,spacedim>::Strategy strategy = cell_diameter_over_24);
```
Comment (bangerth) [inline include/deal.II/numerics/error_estimator.h]: please document the new argument in the documentation of this function (and of all other functions that are not simply referring to this one in their documentation texts).

File: tests/deal.II/error_estimator_02.cc [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+/* Author: Denis Davydov, University of Erlangen-Nuremberg, 2015 */
```
Comment (bangerth) [inline tests/deal.II/error_estimator_02.cc]: add a line here that explains what the test does

File: tests/deal.II/error_estimator_02.cc [hunk 10]
```diff
@@ -0,0 +1,622 @@
...
+#include <deal.II/numerics/error_estimator.h>
+
+// c++
+#include <fstream>
+#include <iostream>
+
+using namespace dealii;
+
+// make sure that the modified Kelly error estimator with
+// strategy = face_diameter_over_twice_max_degree
```
Comment (bangerth) [inline tests/deal.II/error_estimator_02.cc]: or move this long comment to the top so that it's clear that this is what explains what the test does.

**Comment by bangerth** [comment]
Yes, just a couple final touches and then squash.

**Comment by bangerth** [comment]
Oh, and address indentation: https://travis-ci.org/dealii/dealii/builds/67268302

**Comment by davydden** [comment]
done.

**Comment by bangerth** [comment]
@davydden : Would you mind writing a follow-up patch addressing @msteigemann's points? Thanks!

**Pull request merged**